README.md
fbv 1.1 - Very simple framebuffer displayer very useful for embedded splash screens

Modified the fbv-1.0.b source to fix the following:

 - gifs will display all frames

 - using giflib and not libungif

 - re-organize memory allocs and frees to be consistent

 - frame buffer display redraws background for alpha (useful for gifs)

 - add debug configuration option

 - changes to support better cross-compiling

 - fix formating to use hard tabs everywhere


ChangeLog
1.1		2017-08-20		Kyle Farnsworth <kyle@farnsworthtech.com>
      * gifs display all frames
      * re-organize memory allocs and frees to be consistent
      * frame buffer display redraws background for alpha
      * add debug configuration option
      * changes to support better cross-compiling
      * fix formating to use hard tabs everywhere

1.0c     2011-02-09      Marco Cavallini <m.cavallini@koansoftware.com>
      * Added BGR666@24bpp
      * set #define DEFAULT_FRAMEBUFFER "/dev/fb0"

1.0b	2004-09-07
      * unkown modification

0.99    2003-08-24      Mateusz Golicz mtg@elsat.net.pl
      * fixes in transparent and interlaced GIF & PNG support
      * support for using the alpha channel
	  * docs update

0.98    2002-12-05      Nat Ersoz       nat.ersoz@myrio.com
        * fbv now supports the "stride" feature (it works (hopefully)
	  well when a device supports a line length longer than
	  the viewable graphics display)

0.96	2002-01-15	Tomasz Sterna	smoku@jaszczur.org

	* BMP files support
	  (RLE compression and 16bpp not supported yet)

0.95	2001-08-09	Tomasz Sterna	smoku@jaszczur.org

	* dropped autotools
	* added tgt-like configure script

0.94b	2001-02-19	Marcin Zieba    piaskowy@irc.pl
	
	* adding automake support
	* manual page

	2001-02-18	Marcin Zieba	piaskowy@irc.pl
	
	* main.c: added long options support, fixed? segv. Checked, if user supplied filename.

-0.93b	2000-11-23	Tomasz 'smoku' Sterna  <smoku@linux-pl.com>
			Mateusz 'mteg' Golicz  <mtg@elsat.net.pl>
	
	* added this entry just to be fair, they did it all.


bmp.c
/*
    fbv  --  simple image viewer for the linux framebuffer
    Copyright (C) 2002  Tomasz Sterna
	Modified 2017 Kyle Farnsworth (kyle@farnsworthtech.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
#include "config.h"
#ifdef FBV_SUPPORT_BMP
#include "fbv.h"
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#define BMP_TORASTER_OFFSET	10
#define BMP_SIZE_OFFSET		18
#define BMP_BPP_OFFSET		28
#define BMP_RLE_OFFSET		30
#define BMP_COLOR_OFFSET	54

#define fill4B(a)	( ( 4 - ( (a) % 4 ) ) & 0x03)

struct color {
	unsigned char red;
	unsigned char green;
	unsigned char blue;
};

int fh_bmp_id(char *name)
{
	int fd;
	char id[2];
	
	fd = open(name, O_RDONLY);
	if (fd == -1) {
		return(0);
	}
	
	read(fd, id, 2);
	close(fd);
	if ( id[0]=='B' && id[1]=='M' ) {
		return(1);
	}
	return(0);
}

void fetch_pallete(int fd, struct color pallete[], int count)
{
	unsigned char buff[4];
	int i;

	lseek(fd, BMP_COLOR_OFFSET, SEEK_SET);
	for (i=0; i<count; i++) {
		read(fd, buff, 4);
		pallete[i].red = buff[2];
		pallete[i].green = buff[1];
		pallete[i].blue = buff[0];
	}
	return;
}

int fh_bmp_load(char *name,unsigned char **buffer, unsigned char **alpha, int x,int y)
{
	int fd, bpp, raster, i, j, k, skip;
	unsigned char buff[4];
	unsigned char *bp, *wr_buffer;
	struct color pallete[256];

	wr_buffer = (unsigned char*)malloc(x*y*3);
	if (!wr_buffer)
		return FH_ERROR_MEM;
	bp = wr_buffer + x*(y-1)*3;

	fd = open(name, O_RDONLY);
	if (fd == -1) {
		return(FH_ERROR_FILE);
	}

	if (lseek(fd, BMP_TORASTER_OFFSET, SEEK_SET) == -1) {
		free(wr_buffer);
		return(FH_ERROR_FORMAT);
	}
	read(fd, buff, 4);
	raster = buff[0] + (buff[1]<<8) + (buff[2]<<16) + (buff[3]<<24);

	if (lseek(fd, BMP_BPP_OFFSET, SEEK_SET) == -1) {
		free(wr_buffer);
		return(FH_ERROR_FORMAT);
	}
	read(fd, buff, 2);
	bpp = buff[0] + (buff[1]<<8);
	
	switch (bpp){
		case 1: /* monochrome */
			skip = fill4B(x/8+(x%8?1:0));
			lseek(fd, raster, SEEK_SET);
			for (i=0; i<y; i++) {
				for (j=0; j<x/8; j++) {
					read(fd, buff, 1);
					for (k=0; k<8; k++) {
						if (buff[0] & 0x80) {
							*bp++ = 0xff;
							*bp++ = 0xff;
							*bp++ = 0xff;
						} else {
							*bp++ = 0x00;
							*bp++ = 0x00;
							*bp++ = 0x00;
						}
						buff[0] = buff[0]<<1;
					}
					
				}
				if (x%8) {
					read(fd, buff, 1);
					for (k=0; k<x%8; k++) {
						if (buff[0] & 0x80) {
							*bp++ = 0xff;
							*bp++ = 0xff;
							*bp++ = 0xff;
						} else {
							*bp++ = 0x00;
							*bp++ = 0x00;
							*bp++ = 0x00;
						}
						buff[0] = buff[0]<<1;
					}
					
				}
				if (skip) {
					read(fd, buff, skip);
				}
				bp -= x*6; /* backoff 2 lines - x*2 *3 */
			}
			break;
		case 4: /* 4bit palletized */
			skip = fill4B(x/2+x%2);
			fetch_pallete(fd, pallete, 16);
			lseek(fd, raster, SEEK_SET);
			for (i=0; i<y; i++) {
				for (j=0; j<x/2; j++) {
					read(fd, buff, 1);
					buff[1] = buff[0]>>4;
					buff[2] = buff[0] & 0x0f;
					*bp++ = pallete[buff[1]].red;
					*bp++ = pallete[buff[1]].green;
					*bp++ = pallete[buff[1]].blue;
					*bp++ = pallete[buff[2]].red;
					*bp++ = pallete[buff[2]].green;
					*bp++ = pallete[buff[2]].blue;
				}
				if (x%2) {
					read(fd, buff, 1);
					buff[1] = buff[0]>>4;
					*bp++ = pallete[buff[1]].red;
					*bp++ = pallete[buff[1]].green;
					*bp++ = pallete[buff[1]].blue;
				}
				if (skip) {
					read(fd, buff, skip);
				}
				bp -= x*6; /* backoff 2 lines - x*2 *3 */
			}
			break;
		case 8: /* 8bit palletized */
			skip = fill4B(x);
			fetch_pallete(fd, pallete, 256);
			lseek(fd, raster, SEEK_SET);
			for (i=0; i<y; i++) {
				for (j=0; j<x; j++) {
					read(fd, buff, 1);
					*bp++ = pallete[buff[0]].red;
					*bp++ = pallete[buff[0]].green;
					*bp++ = pallete[buff[0]].blue;
				}
				if (skip) {
					read(fd, buff, skip);
				}
				bp -= x*6; /* backoff 2 lines - x*2 *3 */
			}
			break;
		case 16: /* 16bit RGB */
			free(wr_buffer);
			return(FH_ERROR_FORMAT);
			break;
		case 24: /* 24bit RGB */
			skip = fill4B(x*3);
			lseek(fd, raster, SEEK_SET);
			for (i=0; i<y; i++) {
				for (j=0; j<x; j++) {
					read(fd, buff, 3);
					*bp++ = buff[2];
					*bp++ = buff[1];
					*bp++ = buff[0];
				}
				if (skip) {
					read(fd, buff, skip);
				}
				bp -= x*6; /* backoff 2 lines - x*2 *3 */
			}
			break;
		default:
			free(wr_buffer);
			return(FH_ERROR_FORMAT);
	}

	close(fd);
	*buffer = wr_buffer;
	if (alpha)
		*alpha = NULL;
	return(FH_ERROR_OK);
}
int fh_bmp_getsize(char *name,int *x,int *y)
{
	int fd;
	unsigned char size[4];

	fd = open(name, O_RDONLY);
	if (fd == -1) {
		return(FH_ERROR_FILE);
	}
	if (lseek(fd, BMP_SIZE_OFFSET, SEEK_SET) == -1) {
		return(FH_ERROR_FORMAT);
	}
	
	read(fd, size, 4);
	*x = size[0] + (size[1]<<8) + (size[2]<<16) + (size[3]<<24);
//	*x-=1;
	read(fd, size, 4);
	*y = size[0] + (size[1]<<8) + (size[2]<<16) + (size[3]<<24);
	
	close(fd);
	return(FH_ERROR_OK);
}
#endif


fbv.h
/*
    fbv  --  simple image viewer for the linux framebuffer
    Copyright (C) 2000, 2001, 2003  Mateusz Golicz
	Modified 2017 Kyle Farnsworth (kyle@farnsworthtech.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#define FH_ERROR_OK 0
#define FH_ERROR_FILE 1		/* read/access error */
#define FH_ERROR_FORMAT 2	/* file format error */
#define FH_ERROR_MEM 3		/* memory alloc error */

void fb_display(unsigned char *rgbbuff, unsigned char * alpha, int x_size, int y_size, int x_pan, int y_pan, int x_offs, int y_offs, unsigned char **savebuf, int save);
void getCurrentRes(int *x, int *y);

int fh_bmp_id(char *name);
int fh_bmp_load(char *name,unsigned char **buffer, unsigned char **alpha, int x,int y);
int fh_bmp_unload(void);
int fh_bmp_getsize(char *name,int *x,int *y);

int fh_jpeg_id(char *name);
int fh_jpeg_load(char *name,unsigned char **buffer, unsigned char **alpha, int x,int y);
int fh_jpeg_unload(void);
int fh_jpeg_getsize(char *name,int *x,int *y);

int fh_png_id(char *name);
int fh_png_load(char *name,unsigned char **buffer, unsigned char **alpha, int x,int y);
int fh_png_unload(void);
int fh_png_getsize(char *name,int *x,int *y);

int fh_gif_id(char *name);
int fh_gif_load(char *name,unsigned char **buffer, unsigned char **alpha, int x,int y);
int fh_gif_next(unsigned char **buffer, unsigned char **alpha, int x, int y);
int fh_gif_unload(void);
int fh_gif_getsize(char *name,int *x,int *y);
int fh_gif_get_delay(void);
int fh_gif_get_disposal_method(void);
int fh_gif_get_userinput(void);

struct image
{
	int width, height;
	unsigned char *rgb;
	unsigned char *alpha;
	unsigned char *nextrgb;
	unsigned char *nextalpha;
	unsigned char *saved;
};

#ifndef min
#define min(a,b) ((a) < (b) ? (a) : (b))
#endif
#ifndef max
#define max(a,b) ((a) > (b) ? (a) : (b))
#endif
unsigned char * simple_resize(unsigned char * orgin,int ox,int oy,int dx,int dy);
unsigned char * alpha_resize(unsigned char * alpha,int ox,int oy,int dx,int dy);
unsigned char * color_average_resize(unsigned char * orgin,int ox,int oy,int dx,int dy);
unsigned char * rotate(unsigned char *i, int ox, int oy, int rot);
unsigned char * alpha_rotate(unsigned char *i, int ox, int oy, int rot);

#ifdef DEBUG
	extern int debugme;
	#define BUG if (debugme) { fprintf(stderr, "BUG: %s line=%d\n",__FILE__, __LINE__); } 
#else
	#define debugme 0
	#define BUG
#endif



png.c
/*
    fbv  --  simple image viewer for the linux framebuffer
    Copyright (C) 2000, 2001, 2003  Mateusz Golicz
	Modified 2017 Kyle Farnsworth (kyle@farnsworthtech.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "config.h"
#ifdef FBV_SUPPORT_PNG
#include <png.h>
#include "fbv.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#define PNG_BYTES_TO_CHECK 4
#ifndef min
#define min(x,y) ((x) < (y) ? (x) : (y))
#endif

int fh_png_id(char *name)
{
	int fd;
	char id[4];
	fd=open(name,O_RDONLY); if(fd==-1) return(0);
	read(fd,id,4);
	close(fd);
	if(id[1]=='P' && id[2]=='N' && id[3]=='G') return(1);
	return(0);
}
			    

int fh_png_load(char *name,unsigned char **buffer, unsigned char ** alpha,int x,int y)
{
	png_structp png_ptr;
	png_infop info_ptr;
	png_uint_32 width, height;
	int i;
	int bit_depth, color_type, interlace_type;
	int number_passes,pass, trans = 0;
	char *rp;
	png_bytep rptr[2];
	char *fbptr;
	unsigned char *wr_buffer;
	FILE *fh;

	if(!(fh=fopen(name,"rb")))
		return(FH_ERROR_FILE);

	png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING,NULL,NULL,NULL);
	if (png_ptr == NULL)
	{
		return(FH_ERROR_FORMAT);
	}
	info_ptr = png_create_info_struct(png_ptr);
	if (info_ptr == NULL)
	{
		png_destroy_read_struct(&png_ptr, (png_infopp)NULL, (png_infopp)NULL);
		fclose(fh);
		return(FH_ERROR_FORMAT);
	}
	rp=0;
	if (setjmp(png_jmpbuf(png_ptr)))
	{
		png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
		if(rp) free(rp);
		fclose(fh);
		return(FH_ERROR_FORMAT);
	}

	wr_buffer = (unsigned char*)malloc(x*y*3);
	if (!wr_buffer)
		return FH_ERROR_MEM;


	png_init_io(png_ptr,fh);

	png_read_info(png_ptr, info_ptr);
	png_get_IHDR(png_ptr, info_ptr, &width, &height, &bit_depth, &color_type,&interlace_type, NULL, NULL);
	if (color_type == PNG_COLOR_TYPE_PALETTE) png_set_expand(png_ptr); 
	if (bit_depth < 8) png_set_packing(png_ptr);
	if (color_type == PNG_COLOR_TYPE_GRAY || color_type== PNG_COLOR_TYPE_GRAY_ALPHA) png_set_gray_to_rgb(png_ptr);
	if (png_get_valid(png_ptr, info_ptr, PNG_INFO_tRNS))
	{
		trans = 1;
		png_set_tRNS_to_alpha(png_ptr);
	}

	if(bit_depth == 16) png_set_strip_16(png_ptr); 
	number_passes = png_set_interlace_handling(png_ptr);
	png_read_update_info(png_ptr,info_ptr);

	if (color_type == PNG_COLOR_TYPE_GRAY_ALPHA || color_type == PNG_COLOR_TYPE_RGB_ALPHA || trans)
	{
		unsigned char * aptr=NULL;
		
		rp = (char*) malloc(width * 4);
	    rptr[0] = (png_bytep) rp;
		if (alpha)
			*alpha = (unsigned char*) malloc(width * height);
   
	    for (pass = 0; pass < number_passes; pass++)
		{
			fbptr = wr_buffer;
			if (alpha)
				aptr = *alpha;
			
			for(i=0; i<height; i++)
			{
				int n;
				unsigned char *trp = rp;
				
				png_read_rows(png_ptr, rptr, NULL, 1);
				
				for(n = 0; n < width; n++, fbptr += 3, trp += 4)
				{
					memcpy(fbptr, trp, 3);
					if (aptr)
						*(aptr++) = trp[3];
				}
			}
		}
	    free(rp);
	}
	else
	{
	    for (pass = 0; pass < number_passes; pass++)
		{
			fbptr = wr_buffer;
			for(i=0; i<height; i++, fbptr += width*3)
			{
			    rptr[0] = (png_bytep) fbptr;
				png_read_rows(png_ptr, rptr, NULL, 1);
			}
		}
		if (alpha)
			*alpha = NULL;
	}
	png_read_end(png_ptr, info_ptr);
	png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
	fclose(fh);
	*buffer = wr_buffer;
	return(FH_ERROR_OK);
}

int fh_png_getsize(char *name,int *x,int *y)
{
	png_structp png_ptr;
	png_infop info_ptr;
	png_uint_32 width, height;
	int bit_depth, color_type, interlace_type;
	char *rp;
	FILE *fh;

	if(!(fh=fopen(name,"rb"))) return(FH_ERROR_FILE);

	png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING,NULL,NULL,NULL);
	if (png_ptr == NULL) return(FH_ERROR_FORMAT);
	info_ptr = png_create_info_struct(png_ptr);
	if (info_ptr == NULL)
	{
		png_destroy_read_struct(&png_ptr, (png_infopp)NULL, (png_infopp)NULL);
		fclose(fh); return(FH_ERROR_FORMAT);
	}
	rp=0;
	if (setjmp(png_jmpbuf(png_ptr)))
	{
		png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
		if(rp) free(rp);
		fclose(fh); return(FH_ERROR_FORMAT);
	}
   
	png_init_io(png_ptr,fh);
	png_read_info(png_ptr, info_ptr);
	png_get_IHDR(png_ptr, info_ptr, &width, &height, &bit_depth, &color_type,&interlace_type, NULL, NULL);
	png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
	*x=width;
	*y=height;
	fclose(fh);
	return(FH_ERROR_OK);
}
#endif


fb_display.c
/*`
    fbv  --  simple image viewer for the linux framebuffer
    Copyright (C) 2000  Tomasz Sterna
    Copyright (C) 2003  Mateusz Golicz

    Copyright (C) 2011  Marco Cavallini <m.cavallini@koansoftware.com>
    Added BGR666@24bpp
	Modified 2017 Kyle Farnsworth (kyle@farnsworthtech.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <linux/fb.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/mman.h>
#include <asm/types.h>
#include <string.h>
#include <errno.h>
#include "config.h"
/* Public Use Functions:
 *
 * extern void fb_display(unsigned char *rgbbuff,
 *     int x_size, int y_size,
 *     int x_pan, int y_pan,
 *     int x_offs, int y_offs);
 *
 * extern void getCurrentRes(int *x,int *y);
 *
 */


#define DEFAULT_FRAMEBUFFER 	"/dev/fb0"



unsigned short red[256], green[256], blue[256];
struct fb_cmap map332 = {0, 256, red, green, blue, NULL};
unsigned short red_b[256], green_b[256], blue_b[256];
struct fb_cmap map_back = {0, 256, red_b, green_b, blue_b, NULL};


int openFB(const char *name);
void closeFB(int fh);
void getVarScreenInfo(int fh, struct fb_var_screeninfo *var);
void setVarScreenInfo(int fh, struct fb_var_screeninfo *var);
void getFixScreenInfo(int fh, struct fb_fix_screeninfo *fix);
void set332map(int fh);
void* convertRGB2FB(int fh, unsigned char *rgbbuff, unsigned long count, int bpp, int *cpp);
void blit2FB(int fh, void *fbbuff, unsigned char *alpha,
	unsigned int pic_xs, unsigned int pic_ys,
	unsigned int scr_xs, unsigned int scr_ys,
	unsigned int xp, unsigned int yp,
	unsigned int xoffs, unsigned int yoffs,
	int cpp,
	unsigned char **savebuf, int save);

void fb_display(unsigned char *rgbbuff, unsigned char * alpha, int x_size, int y_size, int x_pan, int y_pan, int x_offs, int y_offs, unsigned char **savebuf, int save)
{
    struct fb_var_screeninfo var;
    struct fb_fix_screeninfo fix;
    unsigned short *fbbuff = NULL;
    int fh = -1, bp = 0;
    unsigned long x_stride;
    
    /* get the framebuffer device handle */
    fh = openFB(NULL);
    
    /* read current video mode */
    getVarScreenInfo(fh, &var);
    getFixScreenInfo(fh, &fix);
    
    x_stride = (fix.line_length * 8) / var.bits_per_pixel;
   
    /* correct panning */
    if(x_pan > x_size - x_stride) x_pan = 0;
    if(y_pan > y_size - var.yres) y_pan = 0;
    /* correct offset */
    if(x_offs + x_size > x_stride) x_offs = 0;
    if(y_offs + y_size > var.yres) y_offs = 0;
    
    /* blit buffer 2 fb */
    fbbuff = convertRGB2FB(fh, rgbbuff, x_size * y_size, var.bits_per_pixel, &bp);
#if 0
    blit2FB(fh, fbbuff, alpha, x_size, y_size, x_stride, var.yres, x_pan, y_pan, x_offs, y_offs, bp);
#else
    blit2FB(fh, fbbuff, alpha, x_size, y_size, x_stride, var.yres_virtual, x_pan, y_pan, x_offs, y_offs + var.yoffset, bp, savebuf, save);
#endif
    free(fbbuff);
   
    /* close device */
    closeFB(fh);
}

void getCurrentRes(int *x, int *y)
{
    struct fb_var_screeninfo var;
    int fh = -1;
    fh = openFB(NULL);
    getVarScreenInfo(fh, &var);
    *x = var.xres;
    *y = var.yres;
    closeFB(fh);
}

int openFB(const char *name)
{
    int fh;
    char *dev;

    if(name == NULL){
	dev = getenv("FRAMEBUFFER");
	if(dev) name = dev;
	else name = DEFAULT_FRAMEBUFFER;
    }
    
    if ((fh = open(name, O_RDWR)) == -1){
        fprintf(stderr, "open %s: %s\n", name, strerror(errno));
	exit(1);
    }
    return fh;
}

void closeFB(int fh)
{
    close(fh);
}

void getVarScreenInfo(int fh, struct fb_var_screeninfo *var)
{
    if (ioctl(fh, FBIOGET_VSCREENINFO, var)){
        fprintf(stderr, "ioctl FBIOGET_VSCREENINFO: %s\n", strerror(errno));
	exit(1);
    }
}

void setVarScreenInfo(int fh, struct fb_var_screeninfo *var)
{
    if (ioctl(fh, FBIOPUT_VSCREENINFO, var)){
        fprintf(stderr, "ioctl FBIOPUT_VSCREENINFO: %s\n", strerror(errno));
	exit(1);
    }
}

void getFixScreenInfo(int fh, struct fb_fix_screeninfo *fix)
{
    if (ioctl(fh, FBIOGET_FSCREENINFO, fix)){
        fprintf(stderr, "ioctl FBIOGET_FSCREENINFO: %s\n", strerror(errno));
	exit(1);
    }
}

void make332map(struct fb_cmap *map)
{
	int rs, gs, bs, i;
	int r = 8, g = 8, b = 4;

	map->red = red;
	map->green = green;
	map->blue = blue;

	rs = 256 / (r - 1);
	gs = 256 / (g - 1);
	bs = 256 / (b - 1);
	
	for (i = 0; i < 256; i++) {
		map->red[i]   = (rs * ((i / (g * b)) % r)) * 255;
		map->green[i] = (gs * ((i / b) % g)) * 255;
		map->blue[i]  = (bs * ((i) % b)) * 255;
	}
}

void set8map(int fh, struct fb_cmap *map)
{
    if (ioctl(fh, FBIOPUTCMAP, map) < 0) {
        fprintf(stderr, "Error putting colormap");
        exit(1);
    }
}

void get8map(int fh, struct fb_cmap *map)
{
    if (ioctl(fh, FBIOGETCMAP, map) < 0) {
        fprintf(stderr, "Error getting colormap");
        exit(1);
    }
}

void set332map(int fh)
{
    make332map(&map332);
    set8map(fh, &map332);
}

void blit2FB(int fh, void *fbbuff, unsigned char *alpha,
	unsigned int pic_xs, unsigned int pic_ys,
	unsigned int scr_xs, unsigned int scr_ys,
	unsigned int xp, unsigned int yp,
	unsigned int xoffs, unsigned int yoffs,
	int cpp,
	unsigned char **savebuf, int save)
{
    int i, xc, yc;
	unsigned char *fb;
	
	unsigned char *fbptr;
	unsigned char *imptr;
	unsigned char *saveptr=NULL;
	
    xc = (pic_xs > scr_xs) ? scr_xs : pic_xs;
    yc = (pic_ys > scr_ys) ? scr_ys : pic_ys;

#if 0
	/* if you need to debug */
	printf("-----------------\n") ;
	printf("pic_xs=%d\n", pic_xs) ;
	printf("pic_ys=%d\n", pic_ys) ;
	printf("scr_xs=%d\n", scr_xs) ;
	printf("scr_ys=%d\n", scr_ys) ;
	printf("xp=%d\n", xp) ;
	printf("yp=%d\n", yp) ;
	printf("xoffs=%d\n", xoffs) ;
	printf("yoffs=%d\n", yoffs) ;
	printf("cpp=%d\n", cpp) ;
	printf("xc=%d\n", xc) ;
	printf("yc=%d\n", yc) ;
	printf("-----------------\n") ;
#endif
    
	fb = mmap(NULL, scr_xs * scr_ys * cpp, PROT_WRITE | PROT_READ, MAP_SHARED, fh, 0);
	
	if(fb == MAP_FAILED)
	{
		perror("mmap");
		return;
	}

	if(cpp == 1)
	{
	    get8map(fh, &map_back);
	    set332map(fh);
	}

	fbptr = fb     + (yoffs * scr_xs + xoffs) * cpp;
	imptr = fbbuff + (yp	* pic_xs + xp) * cpp;
	
	if(alpha)
	{
	 	unsigned char * alphaptr;
		int from, to, x;

		if (savebuf)
		{
			if (save)
			{
				unsigned char *sp = malloc(xc * yc * cpp);
				if (sp)
				{
					unsigned char *p=sp, *p2=fbptr;
	    			for(i = 0; i < yc; i++, p2 += scr_xs * cpp, p += xc * cpp)
						memcpy(p, p2, xc * cpp);
					*savebuf = sp;
				}
			}
			saveptr = *savebuf;
		}

		alphaptr = alpha + (yp	* pic_xs + xp);
		
		for(i = 0; i < yc; i++, fbptr += scr_xs * cpp, imptr += pic_xs * cpp, alphaptr += pic_xs)
		{
			if (saveptr)
				memcpy(fbptr, saveptr + (i * pic_xs * cpp), xc * cpp);
			else
				memset(fbptr, 0x00, xc * cpp);
			for(x = 0; x<xc; x++)
			{
				int v;
				
				from = to = -1;
				for(v = x; v<xc; v++)
				{
					if(from == -1)
					{
						if(alphaptr[v] > 0x80) from = v;
					}
					else
					{
						if(alphaptr[v] < 0x80)
						{
							to = v;
							break;
						}
					}
				}
				if(from == -1)
				{
					break;
				}
					
				if(to == -1) to = xc;
				
				memcpy(fbptr + (from * cpp), imptr + (from * cpp), (to - from - 1) * cpp);
				x += to - from - 1;
			}
		}
	}
	else
	    for(i = 0; i < yc; i++, fbptr += scr_xs * cpp, imptr += pic_xs * cpp)
			memcpy(fbptr, imptr, xc * cpp);
		
	if(cpp == 1)
	    set8map(fh, &map_back);
	
	munmap(fb, scr_xs * scr_ys * cpp);
}

inline static unsigned char make8color(unsigned char r, unsigned char g, unsigned char b)
{
    return (
	(((r >> 5) & 7) << 5) |
	(((g >> 5) & 7) << 2) |
	 ((b >> 6) & 3)       );
}

inline static unsigned short make15color(unsigned char r, unsigned char g, unsigned char b)
{
    return (
	(((r >> 3) & 31) << 10) |
	(((g >> 3) & 31) << 5)  |
	 ((b >> 3) & 31)        );
}

inline static unsigned short make16color(unsigned char r, unsigned char g, unsigned char b)
{
    return (
	(((r >> 3) & 31) << 11) |
	(((g >> 2) & 63) << 5)  |
	 ((b >> 3) & 31)        );
}

void* convertRGB2FB(int fh, unsigned char *rgbbuff, unsigned long count, int bpp, int *cpp)
{
    unsigned long i;
    void *fbbuff = NULL;
	u_int8_t  *c_fbbuff;
    u_int16_t *s_fbbuff;
    u_int32_t *i_fbbuff;

    switch(bpp)
    {
	case 8:
	    *cpp = 1;
	    c_fbbuff = (unsigned char *) malloc(count * sizeof(unsigned char));
	    for(i = 0; i < count; i++)
		c_fbbuff[i] = make8color(rgbbuff[i*3], rgbbuff[i*3+1], rgbbuff[i*3+2]);
	    fbbuff = (void *) c_fbbuff;
	    break;
	case 15:
	    *cpp = 2;
	    s_fbbuff = (unsigned short *) malloc(count * sizeof(unsigned short));
	    for(i = 0; i < count ; i++)
		s_fbbuff[i] = make15color(rgbbuff[i*3], rgbbuff[i*3+1], rgbbuff[i*3+2]);
	    fbbuff = (void *) s_fbbuff;
	    break;
	case 16:
	    *cpp = 2;
	    s_fbbuff = (unsigned short *) malloc(count * sizeof(unsigned short));
	    for(i = 0; i < count ; i++)
		s_fbbuff[i] = make16color(rgbbuff[i*3], rgbbuff[i*3+1], rgbbuff[i*3+2]);
	    fbbuff = (void *) s_fbbuff;
	    break;
    case 24:  /* BGR666 */
	    *cpp = 3;
	    i_fbbuff = (unsigned int *) malloc(count * sizeof(unsigned int));
	    for(i = 0; i < count * 3 ; i += 3)
		{   // Skip 24 bit at a time
			c_fbbuff = (u_int8_t  *)i_fbbuff ;
		    c_fbbuff[i + 0] = (u_int8_t)( (rgbbuff[i+2] >> 2) | ((rgbbuff[i+1] & 0x0C) << 4) );
		    c_fbbuff[i + 1] = (u_int8_t)( ((rgbbuff[i+1] & 0xF0) >> 4) | ((rgbbuff[i+0] & 0x3C) << 2) );
		    c_fbbuff[i + 2] = (u_int8_t)( (rgbbuff[i+0] & 0xC0) >> 6 );
		}
	    fbbuff = (void *) i_fbbuff;
	    break;
	case 32:
	    *cpp = 4;
	    i_fbbuff = (unsigned int *) malloc(count * sizeof(unsigned int));
	    for(i = 0; i < count ; i++)
		i_fbbuff[i] = ((rgbbuff[i*3] << 16) & 0xFF0000) |
			    ((rgbbuff[i*3+1] << 8) & 0xFF00) |
			    (rgbbuff[i*3+2] & 0xFF);
	    fbbuff = (void *) i_fbbuff;
	    break;
	default:
	    fprintf(stderr, "Unsupported video mode! You've got: %dbpp\n", bpp);
	    exit(1);
    }
    return fbbuff;
}
